README.md
# A Reverse Proxy for Rack
[![TravisCI](https://secure.travis-ci.org/waterlink/rack-reverse-proxy.svg "Build Status")](http://travis-ci.org/waterlink/rack-reverse-proxy "Build Status")

This is a simple reverse proxy for Rack that pretty heavily rips off Rack Forwarder. It is not meant for production systems (although it may work), as the webserver fronting your app is generally much better at this sort of thing.

## Installation
The gem is available on rubygems.  Assuming you have a recent version of Rubygems you should just be able to install it via:

```
gem install rack-reverse-proxy
```

For your Gemfile use:

```ruby
gem "rack-reverse-proxy", require: "rack/reverse_proxy"
```

## Usage

`Rack::ReverseProxy` should ideally be the very first middleware in your
stack. In a typical use case it is being used to proxy an entirely
different website through your application, so it's unlikely that you will want
any other middleware to modify the requests or responses. The examples below
reflect this.


### Generic Rack app example

```ruby
require 'rack/reverse_proxy'

use Rack::ReverseProxy do
  # Set :preserve_host to true globally (default is true already)
  reverse_proxy_options preserve_host: true

  # Forward the path /test* to http://example.com/test*
  reverse_proxy '/test', 'http://example.com/'

  # Forward the path /foo/* to http://example.com/bar/*
  reverse_proxy /^\/foo(\/.*)$/, 'http://example.com/bar$1', username: 'name', password: 'basic_auth_secret'
end

app = proc do |env|
  [ 200, {'Content-Type' => 'text/plain'}, ["b"] ]
end
run app
```

### Ruby on Rails app example

This example use `config.middleware.insert(0` to ensure that
`Rack::ReverseProxy` is first in the stack. It is possible that
other code in your app (usually in application.rb, development.rb, or production.rb)
will take over this position in the stack. To ensure
that this is not the case, view the stack by running `rails middleware`. You should see
`Rack::ReverseProxy` at the top. Note that
the middleware stack will likely differ slightly in each environment. All that said, it's a pretty
safe bet to put the below code into application.rb.

```ruby
# config/application.rb
config.middleware.insert(0, Rack::ReverseProxy) do
  reverse_proxy_options preserve_host: true
  reverse_proxy '/wiki', 'http://wiki.example.com/'
end
```

### Rules

As seen in the Rack example above, `reverse_proxy` can be invoked multiple times with
different rules, which will be commulatively added.

Rules can be a regex or a string. If a regex is used, you can use the subcaptures in your forwarding url by denoting them with a `$`.

Right now if more than one rule matches any given route, it throws an exception for an ambiguous match.  This will probably change later. If no match is found, the call is forwarded to your application.


### Options

#### Terminology

<dl>
  <dt>Source Request</dt>
  <dd>The request sent from the end user's browser to the rack stack which is running rack-reverse-proxy.</dd>
  <dt>Target Request</dt>
  <dd>The request sent from rack-reverse-proxy to the server which will produce the response.</dd>
</dl>

`reverse_proxy_options` sets global options for all reverse proxies. Available options are:

* `:preserve_host` This determines which Host header is used in the target request. If set to `true`, the target
  request's host (specified in the `reverse_proxy` invocation) will be used. If set to `false`, the source request's
  Host header will be re-used for the target request. Default: `true`.
* `:username` username for basic auth
* `:password` password for basic auth
* `:matching` is a global only option, if set to :first the first matched url will be requested (no ambigous error). Default: :all.
* `:timeout` seconds to timout the requests
* `:force_ssl` redirects to ssl version, if not already using it (requires `:replace_response_host`). Default: false.
* `:verify_mode` the `OpenSSL::SSL` verify mode passed to Net::HTTP. Default: `OpenSSL::SSL::VERIFY_PEER`.
* `:x_forwarded_headers` sets up proper `X-Forwarded-*` headers. Default: true.
* `:stripped_headers` Array of headers that should be stripped before forwarding reqeust. Default: nil.
  e.g. `stripped_headers: ["Accept-Encoding", "Foo-Bar"]`

If `reverse_proxy_options` is invoked multiple times, the invocations will have a commulative effect,
only overwritting the values which they specify. Example of how this could be useful:

```ruby
config.middleware.insert(0, Rack::ReverseProxy) do
  reverse_proxy_options preserve_host: false
  if Rails.env.production? or Rails.env.staging?
    reverse_proxy_options force_ssl: true, replace_response_host: true
  end
  reverse_proxy /^\/blog(\/?.*)$/, 'http://blog.example.com/blog$1'
end
```

## Note on Patches/Pull Requests
* Fork the project.
* Make your feature addition or bug fix.
* Add tests for it. This is important so I don't break it in a
  future version unintentionally.
* Commit, do not mess with rakefile, version, or history.
  (if you want to have your own version, that is fine but bump version in a commit by itself I can ignore when I pull)
* Send me a pull request. Bonus points for topic branches.

## Contributors

- Jon Swope, creator
- Oleksii Fedorov, maintainer


CHANGELOG.md
# Changelog

## 1.0.0 (UNRELEASED)

- Breaking Change: Never modify Location headers that are only paths without hosts. [John Bachir](https://github.com/jjb) [#46](https://github.com/waterlink/rack-reverse-proxy/pull/46)
- Breaking Change: Previously, the Accept-Encoding header was stripped by default, unless the
  `preserve_encoding` option was set to true. Now, no headers are stripped by default, and an array
  of headers that should be stripped can be specified with the `stripped_headers` option. [John Bachir](https://github.com/jjb) [#52](https://github.com/waterlink/rack-reverse-proxy/pull/52)
- Breaking Change: Previously, rack-reverse-proxy had the behavior/bug that when reverse_proxy_options
  was invoked, all options that weren't set in the invokation would be set to nil. Now, those options will remain set at their default values - [Krzysztof Knapik](https://github.com/knapo) [#37](https://github.com/waterlink/rack-reverse-proxy/pull/37) and [John Bachir](https://github.com/jjb) [#47](https://github.com/waterlink/rack-reverse-proxy/pull/47)
- Breaking Change: Previously, when invoking reverse_proxy_options multiple times, only the
  final invocation would have any effect. Now, the invocations will have a commulative effect.
  [John Bachir](https://github.com/jjb) [#47](https://github.com/waterlink/rack-reverse-proxy/pull/47)
- Bugfix: Fix rack response body for https redirects [John Bachir](https://github.com/jjb) [#43](https://github.com/waterlink/rack-reverse-proxy/pull/43)
- Documentation: It is now suggested that the middleware be first in your app's stack unless you have a specific reason to do otherwise [John Bachir](https://github.com/jjb) [#49](https://github.com/waterlink/rack-reverse-proxy/pull/49)


## 0.12.0

- Enhancement: Set "X-Forwarded-Proto" header to the proxying scheme. [Motonobu Kuryu](https://github.com/arc279) [#32](https://github.com/waterlink/rack-reverse-proxy/pull/32)
- Bugfix: Upgrade to a version of rack-proxy with the bug fix for the unclosed network resources. [John Bachir](https://github.com/jjb) [#45](https://github.com/waterlink/rack-reverse-proxy/pull/45)

## 0.11.0

- Breaking Change: Rename option x_forwarded_host option to x_forwarded_headers, as it controls both X-Forwarded-Port and X-Forwarded-Host - [Aurelien Derouineau](https://github.com/aderouineau) [#26](https://github.com/waterlink/rack-reverse-proxy/pull/26)
- Breaking Change: Strip Accept-Encoding header before forwarding request. [Max Gulyaev](https://github.com/maxilev) [#27](https://github.com/waterlink/rack-reverse-proxy/pull/27)

## 0.10.0

- Feature: `options[:verify_mode]` to set SSL verification mode. - [Marv Cool](https://github.com/MrMarvin) [#24](https://github.com/waterlink/rack-reverse-proxy/pull/24) and [#25](https://github.com/waterlink/rack-reverse-proxy/pull/25)

## 0.9.1

- Enhancement: Remove `Status` key from response headers as per Rack protocol (see [rack/lint](https://github.com/rack/rack/blob/master/lib/rack/lint.rb#L639)) - [Jan Raasch](https://github.com/janraasch) [#7](https://github.com/waterlink/rack-reverse-proxy/pull/7)

## 0.9.0

- Bugfix: Timeout option matches the documentation - [Paul Hepworth](https://github.com/peppyheppy)
- Ruby 1.8 compatibility - [anujdas](https://github.com/anujdas)
- Bugfix: Omit port in host header for default ports (80, 443), so that it doesn't break some web servers, like "Apache Coyote" - [Peter Suschlik](https://github.com/splattael)
- Bugfix: Don't drop source request's port in response's location header - [Eric Koslow](https://github.com/ekosz)
- Bugfix: Capitalize headers correctly to prevent duplicate headers when used together with other proxies - [Eric Koslow](https://github.com/ekosz)
- Bugfix: Normalize headers from HttpStreamingResponse in order not to break other middlewares - [Jan Raasch](https://github.com/janraasch)


lib/rack_reverse_proxy/roundtrip.rb
require "rack_reverse_proxy/response_builder"

module RackReverseProxy
  # FIXME: Enable them and fix issues during refactoring
  # rubocop:disable Metrics/ClassLength

  # RoundTrip represents one request-response made by rack-reverse-proxy
  # middleware.
  class RoundTrip
    def initialize(app, env, global_options, rules, response_builder_klass = ResponseBuilder)
      @app = app
      @env = env
      @global_options = global_options
      @rules = rules
      @response_builder_klass = response_builder_klass
    end

    def call
      return app.call(env) if rule.nil?
      return proxy_with_newrelic if new_relic?
      proxy
    end

    private

    attr_reader :app, :env, :global_options, :rules, :response_builder_klass

    def new_relic?
      global_options[:newrelic_instrumentation]
    end

    def proxy_with_newrelic
      perform_action_with_newrelic_trace(:name => action_name, :request => source_request) do
        proxy
      end
    end

    def action_name
      "#{action_path}/#{source_request.request_method}"
    end

    def action_path
      # Rack::ReverseProxy/foo/bar#GET
      source_request.path.gsub(%r{/\d+}, "/:id").gsub(%r{^/}, "")
    end

    def uri
      return @_uri if defined?(@_uri)
      @_uri = rule.get_uri(path, env, headers, source_request)
    end

    def options
      @_options ||= global_options.dup.merge(rule.options)
    end

    def https_redirect
      rewrite_uri(uri, source_request)
      uri.scheme = "https"
      [301, { "Location" => uri.to_s }, [""]]
    end

    def need_https_redirect?
      options[:force_ssl] &&
        options[:replace_response_host] &&
        source_request.scheme == "http"
    end

    def target_request
      @_target_request ||= build_target_request
    end

    def target_request_headers
      @_target_request_headers ||= headers
    end

    def build_target_request
      Net::HTTP.const_get(
        source_request.request_method.capitalize
      ).new(uri.request_uri)
    end

    def preserve_host
      return unless options[:preserve_host]
      target_request_headers["HOST"] = host_header
    end

    def strip_headers
      return unless options[:stripped_headers]
      options[:stripped_headers].each do |header|
        target_request_headers.delete(header)
      end
    end

    def host_header
      return uri.host if uri.port == uri.default_port
      "#{uri.host}:#{uri.port}"
    end

    def set_forwarded_headers
      return unless options[:x_forwarded_headers]
      target_request_headers["X-Forwarded-Host"] = source_request.host
      target_request_headers["X-Forwarded-Port"] = source_request.port.to_s
      target_request_headers["X-Forwarded-Proto"] = source_request.scheme
    end

    def initialize_http_header
      target_request.initialize_http_header(target_request_headers)
    end

    def set_basic_auth
      return unless need_basic_auth?
      target_request.basic_auth(options[:username], options[:password])
    end

    def need_basic_auth?
      options[:username] && options[:password]
    end

    def setup_body
      return unless can_have_body? && body?
      source_request.body.rewind
      target_request.body_stream = source_request.body
    end

    def can_have_body?
      target_request.request_body_permitted?
    end

    def body?
      source_request.body
    end

    def set_content_length
      target_request.content_length = source_request.content_length || 0
    end

    def set_content_type
      return unless content_type?
      target_request.content_type = source_request.content_type
    end

    def content_type?
      source_request.content_type
    end

    def target_response
      @_target_response ||= response_builder_klass.new(
        target_request,
        uri,
        options
      ).fetch
    end

    def response_headers
      @_response_headers ||= build_response_headers
    end

    def build_response_headers
      ["Transfer-Encoding", "Status"].inject(rack_response_headers) do |acc, header|
        acc.delete(header)
        acc
      end
    end

    def rack_response_headers
      Rack::Utils::HeaderHash.new(
        Rack::Proxy.normalize_headers(
          format_headers(target_response.headers)
        )
      )
    rescue Errno::ECONNREFUSED => e
      @_target_response = Struct.new(:status, :body).new(502, [])
      {}
    end

    def replace_location_header
      return unless need_replace_location?
      rewrite_uri(response_location, source_request)
      response_headers["Location"] = response_location.to_s
    end

    def response_location
      @_response_location ||= URI(response_headers["Location"])
    end

    def need_replace_location?
      response_headers["Location"] && options[:replace_response_host] && response_location.host
    end

    def setup_request
      preserve_host
      strip_headers
      set_forwarded_headers
      initialize_http_header
      set_basic_auth
      setup_body
      set_content_length
      set_content_type
    end

    def setup_response_headers
      replace_location_header
    end

    def rack_response
      [target_response.status, response_headers, target_response.body]
    end

    def proxy
      return app.call(env) if uri.nil?
      return https_redirect if need_https_redirect?

      setup_request
      setup_response_headers

      transform_response(rack_response)
    end

    def transform_response(response)
      rule.transform(path, env, response, uri, headers, source_request)
    end

    def format_headers(headers)
      headers.inject({}) do |acc, (key, val)|
        formated_key = key.split("-").map(&:capitalize).join("-")
        acc[formated_key] = Array(val)
        acc
      end
    end

    def request_default_port?(req)
      [["http", 80], ["https", 443]].include?([req.scheme, req.port])
    end

    def rewrite_uri(uri, original_req)
      uri.scheme = original_req.scheme
      uri.host   = original_req.host
      uri.port   = original_req.port unless request_default_port?(original_req)
    end

    def source_request
      @_source_request ||= Rack::Request.new(env)
    end

    def rule
      return @_rule if defined?(@_rule)
      @_rule = find_rule
    end

    def find_rule
      return if matches.empty?
      non_ambiguous_match
      matches.first
    end

    def path
      @_path ||= source_request.fullpath
    end

    def headers
      Rack::Proxy.extract_http_request_headers(source_request.env)
    end

    def matches
      @_matches ||= rules.select do |rule|
        rule.proxy?(path, headers, source_request)
      end
    end

    def non_ambiguous_match
      return unless ambiguous_match?
      raise Errors::AmbiguousMatch.new(path, matches)
    end

    def ambiguous_match?
      matches.length > 1 && global_options[:matching] != :first
    end
  end
end
